### ISSUE TITLE
Zero implicit gradients when 

### ISSUE THREAD

**Issue by itk22** [patrick-kidger/optimistix#54]
Hi @patrick-kidger and @packquickly,

I was trying to implement the following meta-learning example from jax-opt in optimistix: [Few-shot Adaptation with Model Agnostic Meta-Learning ](https://jaxopt.github.io/stable/notebooks/implicit_diff/maml.html). However, I ran into an issue with implicit differentiation through the inner loop. The below example runs well when using `optx.RecursiveCheckpointAdjoint` but when I try to recreate the iMAML setup by putting  `optx.ImplicitAdjoint` with a CG solver with 20 steps, all the meta-gradients are zero, and the meta-optimiser doesn't change at all in the training. Could you please help me identify the issue?

Here is the code that I have so far for your reference (please excuse the verbosity):

```
# Inner-loop optimizer in Optimistix
inner_opt = optx.GradientDescent(learning_rate=cfg.inner_lr,
                                 rtol=1e-12,
                                 atol=1e-12)

# Outer-loop (meta) optimizer in pure optax
meta_opt = optax.adam(outer_lr)

# Create the prediction model
model = create_simple_mlp_model()

# Partition the model - parameters  + state
params, state = eqx.partition(model, eqx.is_inexact_array)

def apply_model(params, state, X):
    model = eqx.combine(params, state)
    return jax.vmap(model)(X)

def mse_loss_fn(X, Y, params):
    pred = apply_model(params, state, X)
    return jnp.square(Y - pred).mean()

def inner_loss_fn(params, args):
    outer_params, sample = args
    # Evaluate the MSE loss on a sample input
    X, Y = sample
    pred = apply_model(params, state, X)
    mse = mse_loss_fn(X, Y, params)
    # Regularizing term
    diff = tree_add_scalar_mul(outer_params, -1, params)
    reg = (regularization_strength / 2) * tree_l2_norm(diff)

    return mse + reg

def outer_loss_fn(params, args):
    sample = args
    # Outer loss function returns the mse without regularization
    X, Y = sample
    pred = apply_model(params, state, X)
    mse = mse_loss_fn(X, Y, params)
    return mse

def single_task_rollout(meta_params, sample, inner_steps):
    # Recursive checkpoint adjoint - works for now
    sol = optx.minimise(
        inner_loss_fn,
        inner_opt,
        y0=meta_params,
        args=(jax.lax.stop_gradient(meta_params), sample),
        max_steps=inner_steps,
        throw=False,
    )

    # Return the adapted params
    return sol.value

def single_task_loss_and_grad(meta_params, train_sample, test_sample):
    def task_rollout_and_eval(meta_params):
        adapted_params = single_task_rollout(meta_params, train_sample,
                                             inner_steps)
        outer_loss = outer_loss_fn(adapted_params, test_sample)
        return outer_loss

    meta_loss, meta_grad = jax.value_and_grad(task_rollout_and_eval)(
        meta_params)

    return meta_loss, meta_grad

def multi_task_loss_and_grad(meta_params, batch):
    train_batch, test_batch = batch
    meta_losses, meta_grads = jax.vmap(single_task_loss_and_grad,
                                       in_axes=(None, 0, 0))(meta_params,
                                                             train_batch,
                                                             test_batch)

    # Average the meta-losses and gradients across the batch
    meta_loss = tree_mean(meta_losses, axis=0)
    meta_grad = tree_mean(meta_grads, axis=0)

    # Return the meta-loss twice - once as aux
    return meta_loss, meta_grad

@jax.jit
def outer_step_scan_fn(carry, _):
    # Written to be compatible with jax.lax.scan
    meta_params, meta_opt_state, key = carry
    # Split the key
    key, _ = jr.split(key, 2)
    # Generate a batch of tasks
    train_batch, test_batch = generate_sin_tasks(batch_size,
                                                 n_points,
                                                 key=key)
    # Get current meta loss and gradients
    meta_grad, meta_loss = multi_task_loss_and_grad(meta_params,
                                                    (train_batch,
                                                     test_batch))

    # Calculate the update
    updates, meta_opt_state = meta_opt.update(meta_grad, meta_opt_state)
    # Apply the update
    meta_params = eqx.apply_updates(meta_params, updates)

    # Return the updated meta-parameters and key, and corresponding losses
    return (meta_params, meta_opt_state, key), meta_loss

###########################################################################
# Training loop
###########################################################################
# Initialize the random key
key = jr.PRNGKey(2)

# Initialize the optimizer state
meta_opt_state = meta_opt.init(params)

# Each outer batch should have its own key - how can we ensure this?
logger.info(f"Starting a meta-training loop with {outer_steps} outer "
            f"and {inner_steps} inner steps.")

t0 = time.time()

(meta_params, meta_opt_state,
 key), meta_losses = jax.lax.scan(outer_step_scan_fn,
                                  init=(params, meta_opt_state, key),
                                  xs=None,
                                  length=outer_steps)
t1 = time.time()
logger.info(f"Finished in {t1 - t0:.2f}s.")
```

**Comment by patrick-kidger** [comment]
Hey there! Thanks for the issue.i

Would you be able to condense your code down to a single MWE? (Preferably around 20 lines of code.) For example we probably don't need the details of your training loop, the fact that it's batched, etc. Moreover I'm afraid this code won't run -- if nothing else, it currently doesn't have any import statements.

**Comment by itk22** [comment]
Hi @patrick-kidger,

I updated the original post with a condensed MWE. In the above code, when I use `optx.RecursiveCheckpointAdjoint()`, I am able to recover the correct gradients. However, when I use `optx.ImplicitAdjoint` with a solver specified as CG, the gradients are all exactly zero.  To be fair, I was not expecting this to work out of the box because 1. `adapt_fn` does not find the exact solution to the inner optimization problem, 2. even for a small network, this seems to be a rather difficult calculation. However, the jax-opt example I shared above indicates that the gradients can be calculated correctly in a similar scenario using CG:
 
<img src='https://github.com/patrick-kidger/optimistix/assets/112338249/9daa7551-da26-4839-91a2-f87033877e65.png' width='300'>

Because of this, I started wondering if there is a fundamental difference in how implicit adjoints are calculated in the two packages. My instinct is that the mismatch might have to do with handling of higher-order terms but I am curious to hear your opinion and whether it is something that can be quickly patched.

**Comment by patrick-kidger** [comment]
So in a minimisation problem, the solution stays constant as you peturb the initial parameters. Regardless of where you start, you should expect to converge to the same solution! So in fact a zero gradient is what is expected. (Imagine finding `argmin_x x^2`. It doesn't matter whether you start at `x=1` or `x=1.1`; either way your output will be `x=0`.)

The fact that you get a nonzero gradient via `RecursiveCheckpointAdjoint` will be because of the fact that you are taking so few steps that you are not actually converging to the minima at all. (In the above example, you might only converge as far as `x=0.5` or `x=0.6`.) So I think for a meta-learning use-case, then probably `RecursiveCheckpointAdjoint` is actually the correct thing to be doing!

The fact that JAXopt appears to do otherwise is possibly a bug in JAXopt. (?)

That aside, some comments on your implementation:
- I would recommend against a `CG` solver if you can help it. This is a fairly numerically inaccurate / unstable solver.
- Other than the discussion above, you also appear to be computing a gradient through `sol.aux`. In its current meaning this is the `aux` from the final step of the solve. This means gradients through this are actually not defined mathematically, since this is an internal detail of the solver and unrelated to the implicit function theorem! So (a) don't try to use this in gradient calculations when using `ImplicitAdjoint`, but also (b) I can still totally see that this is a footgun without guardrails, and you've prompted us to think whether there's a way to adjust this into something better.

**Comment by itk22** [comment]
Hi Patrick,

Thank you for your thorough response. It is true that the solution to a minimisation problem is independent of the initial parameters and should lead to zero gradients. As you noted, the gradients from RecursiveCheckpointAdjoint are non-zero in the above MWE because we take very few optimisation steps. I set that number low to emulate a typical bi-level meta-learning setup where, in the inner loops, we do not fully optimise the model parameters for each task but rather take just a few steps of optimisation. This is because the goal is not to find the true optimum for any single task but rather to optimise the initialisation such that the model can quickly adapt to related tasks. In this case, the gradients through the inner loop, must be non-zero for the initialisation to evolve across outer loop iterations. Also, iMAML has a special regularising term to ensure meta-gradients are not non-zero for larger number of inner steps.

So, there is no bug in JAXopt, and it only appeared so because of my incomplete explanation. The iMAML paper that I am following uses CG because it avoids forming a Hessian matrix. It also seems that CG-like iterative solvers are used quite extensively within JAXopt  - as far as I understand this again has to do with their matrix-free nature. Optimistix, on the other hand seems to be using direct solvers as a default.
